### PR TITLE
Add API support for rendering a rich diff for Copilot

### PR DESCRIPTION
**Pull request by AaronFriel**
Reviewers should note this PR is in two commits.

Updates the progress display to support deterministic output without rendering each in-progress frame, called "plain" output in the tests below.

The CLI should now render tree views more consistently: implementing this test required fixing a non-deterministic ordering bug that affected all tree renders. This bug was caused by iterating over a map to determine the row order, however Go does not guarantee a consistent order when iterating over a map. Now insertions into the tree are ordered by URN, which should be deterministic.

Benchmarks are added to measure timing and memory usage of progress displays, as we intend to use this rendering mode with Pulumi Copilot.

On this laptop the slowest test, `template-body.json`, has its results reported below.

The new "plain" output is the fastest by an order of magnitude - and this scaling factor improves as the size of the test grows. While "plain" here is 15 to 20x faster than the other modes, in shorter tests it is only 2 to 4x faster. Peak heap usage and allocations are also reduced by rendering only the last frame, though not as dramatically.

```
goos: linux
goarch: amd64
pkg: github.com/pulumi/pulumi/pkg/v3/backend/display
cpu: 11th Gen Intel(R) Core(TM) i7-11370H @ 3.30GHz
BenchmarkProgressEvents
BenchmarkProgressEvents/template-body.jsoninteractive/80x24/raw-8         12	  91473186 ns/op	11745682 B/op	  223168 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/80x24/cooked-8       8	 134639710 ns/op	16078474 B/op	  269443 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/80x24/plain-8      152	   7999863 ns/op	 2952236 B/op	   52725 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/100x80/raw-8         8	 142099499 ns/op	14636003 B/op	  239221 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/100x80/cooked-8      7	 147865799 ns/op	17260515 B/op	  275864 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/100x80/plain-8     147	   7874586 ns/op	 2960121 B/op	   52782 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/200x80/raw-8         8	 128632367 ns/op	14650076 B/op	  239220 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/200x80/cooked-8      7	 149342244 ns/op	18052901 B/op	  284310 allocs/op
BenchmarkProgressEvents/template-body.jsoninteractive/200x80/plain-8     144	   7999831 ns/op	 2962243 B/op	   52784 allocs/op
BenchmarkProgressEvents/template-body.jsonnon-interactive-8              178	   6662351 ns/op	 2877507 B/op	   51378 allocs/op
```

Also adds a "ShowResourceChanges" option to the progress display, which when set adds a section showing resource diffs. The updated test data shows examples. This combines the tree view with more detailed information about the changes made, albeit more compactly. This change is intended to be consumed by Pulumi Copilot, and so is not enabled for the CLI at this time.

Each commit of this PR implements these features separately.

### CODE REVIEW

File: pkg/backend/display/progress_test.go [hunk 1]
```diff
@@ -226,6 +243,65 @@ func TestProgressEvents(t *testing.T) {
 	}
 }
 
+func BenchmarkProgressEvents(t *testing.B) {
+	t.Setenv("TERM", "vt102")
```
Comment (arturl) [inline pkg/backend/display/progress_test.go:251]: Why is this necessary in this test and not others?

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
+View in Browser (Ctrl+O): link
```
Comment (arturl) [inline pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:1]: I know you did not introduce this style of testing but I find this "overfitting" - such tests make it hard to introduce unrelated changes. For example, adding `askCopilot` output will require changes in all of these files.

**Review by arturl — APPROVED** [review]
LGTM, see suggestions

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        [urn=urn:pulumi:dev::test::pulumi-python:dynamic:Resource::test-x]
+
+Diagnostics:
+  pulumi:pulumi:Stack (test-dev):
+    Configuration:
+
+Resources:
+    1 unchanged
+
+Duration: 8s
```
Comment (arturl) [inline pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:26]: Re: `8s` - what happens to this test when run on a different hardware?

Comment (AaronFriel) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:26]: The timestamps are calculated from the events:

https://github.com/pulumi/pulumi/blob/8baa8386f194849360b524e091b641db3c04f2f0/pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json#L1-L9

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt (hunk 2 above)
Comment (AaronFriel) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:1]: I think we prefer it because otherwise it is very easy for us to produce incoherent text output, as testing each portion of the rendering in isolation doesn't give a complete picture of what the user sees.

The PR adding askCopilot required a flag, though, but it should probably be "ShowLinkToCopilot" and not "SuppressLinkToCopilot"?

File: pkg/backend/display/progress_test.go (hunk 1 above)
Comment (AaronFriel) [reply pkg/backend/display/progress_test.go:251]: It's necessary for both tests that run progress tests for consistency across machines, as the rendering is TERM specific.

File: pkg/backend/display/progress.go [hunk 4]
```diff
@@ -1463,3 +1427,38 @@ func (display *ProgressDisplay) getStepInProgressDescription(step engine.StepEve
 	}
 	return deploy.ColorProgress(op) + getDescription() + colors.Reset
 }
+
+// toSortedResourceRows sorts a map of resources by their URN and returns the sorted rows.
+func toSortedResourceRows(rowMap map[resource.URN]ResourceRow, deterministicOutput bool) []ResourceRow {
```
[the remark this answers is not in the document]
Comment (AaronFriel) [reply pkg/backend/display/progress.go:1485]: Renaming to `toResourceRows`, with the flag controlling whether it's sorted

File: pkg/backend/display/progress_test.go (hunk 1 above)
Comment (arturl) [reply pkg/backend/display/progress_test.go:251]: OK. Is it not needed in other tests?

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt (hunk 2 above)
Comment (arturl) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:1]: Re: flag name: I followed the pattern of Suppress* names (SuppressOutputs, SuppressPermalink, SuppressTimings, SuppressProgress). All of them are ON by default unless explicitly suppressed. We want that for Copilot links as well.

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt (hunk 3 above)
Comment (arturl) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:26]: Is 8s part of test validation? E.g. if the test takes 9s on a different machine, will the test fail? I supposed not, but I don't see how this number is ignored.

Comment (AaronFriel) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:26]: No, it will always print 8s as long as it's the difference between the timestamps in line 0 and line 9, I believe.

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt (hunk 2 above)
Comment (AaronFriel) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:1]: I suspect not - as we only want to show Copilot when the backend is `api.pulumi.com`?

File: pkg/backend/display/progress_test.go (hunk 1 above)
Comment (AaronFriel) [reply pkg/backend/display/progress_test.go:178]: No, they don't rely on this.

Comment (arturl) [reply pkg/backend/display/progress_test.go:178]: OK!

File: pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt (hunk 2 above)
Comment (pgavlin) [reply pkg/backend/display/testdata/not-truncated/failed-stacks-dont-rewrite-operations.json.interactive-100x80-plain.stdout.txt:1]: > I think we prefer it because otherwise it is very easy for us to produce incoherent text output, as testing each portion of the rendering in isolation doesn't give a complete picture of what the user sees.

Yeah this style is very intentional. It is meant to make changes to the output clear and obvious. (though admittedly the tagged output can be hard to parse).

File: pkg/backend/display/tree.go [hunk 5]
```diff
@@ -189,7 +189,7 @@ func (r *treeRenderer) markDirty() {
 	}
 
 	r.dirty = true
-	if r.opts.DeterministicOutput {
+	if r.opts.DeterministicOutput && r.opts.RenderOnDirty {
```
Comment (pgavlin) [inline pkg/backend/display/tree.go:192]: is this extra flag meant to suppress output until the final frame?

Comment (AaronFriel) [reply pkg/backend/display/tree.go:192]: Yeah
